### PR TITLE
allow patch to handle multiple types

### PR DESCRIPTION
**Pull request by deads2k**
Right now `kubectl patch` only supports strategic merge patch, but its easier to express some kinds of changes using different patch types.  For instance, updates to positional arrays are very easy with a `json` patch type.  This adds the ability to choose a patch type.

@kubernetes/kubectl @YannMoisan

### CODE REVIEW

**Comment by kargakis** [comment]
Add an example and run the update-docs script

**Comment by bgrant0607** [comment]
cc @kubernetes/kubectl

File: pkg/kubectl/cmd/patch.go [hunk 1]
```diff
@@ -68,6 +72,7 @@ func NewCmdPatch(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 		},
 	}
 	cmd.Flags().StringP("patch", "p", "", "The patch to be applied to the resource JSON file.")
+	cmd.Flags().String("type", "strategic", fmt.Sprintf("The type of patch being provided; one of %v", sets.StringKeySet(patchTypes).List()))
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/patch.go]: Please move this below the MarkFlagRequired call.

**Comment by bgrant0607** [comment]
And a test, please.

**Comment by bgrant0607** [comment]
To be more clear: Feature and flag name are ok. Needs example and test.

File: pkg/kubectl/cmd/patch.go [hunk 2]
```diff
@@ -83,6 +88,18 @@ func RunPatch(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []stri
 		return err
 	}
 
+	patchType := api.StrategicMergePatchType
+	patchTypeString := strings.ToLower(cmdutil.GetFlagString(cmd, "type"))
+	if len(patchTypeString) == 0 {
+		patchType = api.StrategicMergePatchType
+	} else {
```
Comment (ghodss) [inline pkg/kubectl/cmd/patch.go]: Can you just make this line `if len(patchTypeString) != 0 {` and delete the two lines above it?

**Comment by ghodss** [comment]
:+1:

**Comment by jackgr** [comment]
:+1:

**Comment by deads2k** [comment]
Example added, tested in test-cmd.sh.

**Comment by kargakis** [comment]
LGTM

**Comment by deads2k** [comment]
@bgrant0607 approved the concept and flagname: https://github.com/kubernetes/kubernetes/pull/21099#issuecomment-183095927 

@derekwaynecarr pick the up the implementation review?  It's small.

**Comment by derekwaynecarr** [comment]
LGTM

**Pull request merged**